Assets/_BForBoss/_Character/Scripts/PlayerWallRunBehaviour.cs

```diff
@@ -0,0 +1,232 @@
+                else Debug.DrawRay(ChildTransform.position, dir * _wallMaxDistance, Color.red);
+            }
+
+            if (GetSmallestRaycastHit(_hits, out RaycastHit hit))
+            {
+                WallRun(hit);
+            }
+            else if (_isWallRunning) StopWallRunning();
+        }
+
+        public void OnJumped(ref int _jumpCount)
+        {
+            _currentJumpDuration = 0f;
+            if (_isWallRunning)
+            {
+                StopWallRunning();
+                _jumpCount = 0;
+            }
+        }
+
+        public void OnLanded()
+        {
+            _currentJumpDuration = 0f;
+            if (_isWallRunning)
+            {
+                StopWallRunning();
+            }
+        }
+
+        public bool CanJump()
+        {
+            return _isWallRunning; // can always jump out of a wall run
+        }
+
+        public void OnWallRunning()
+        {
+            if (!_isWallRunning)
+            {
+                _timeSinceWallDetach += Time.deltaTime;
+                return;
+            }
+
+            if (_movementInput().sqrMagnitude <= 0)
+            {
+                return;
+            }
+            _timeSinceWallAttach += Time.deltaTime;
+            var velocity = _baseCharacter.GetVelocity();
+            var alongWall = ChildTransform.TransformDirection(Vector3.forward).normalized;
+            velocity = velocity.dot(alongWall) * alongWall;
+            velocity += Vector3.down * _wallGravityDownForce * Time.deltaTime;
+            _baseCharacter.SetVelocity(velocity);
+        }
+
+        public bool CalcJumpVelocity(out Vector3 velocity)
+        {
+            velocity = Vector3.zero;
+            if (_isWallRunning)
+            {
+                velocity = _lastWallRunNormal * _baseCharacter.jumpImpulse + Vector3.up;
+            }
+            return _isWallRunning;
+        }
+
+        private void HandleEyePivotRotation()
+        {
+            var rotation = _fpsCharacter.eyePivot.localEulerAngles;
+            rotation.z = GetCameraRoll();
+            _fpsCharacter.eyePivot.localEulerAngles = rotation;
+        }
+
+        private void WallRun(RaycastHit wall)
+        {
+            _lastWallRunNormal = wall.normal;
+            _lastWallRunPosition = wall.point;
+            if (!_isWallRunning)
+            {
+                _isWallRunning = true;
+                _baseMaxSpeed = _baseCharacter.maxWalkSpeed;
+                _baseCharacter.maxWalkSpeed *= _speedMultiplier;
+                _timeSinceWallAttach = 0f;
+                _timeSinceWallDetach = 0f;
+            }
+        }
+
+        private bool CanWallRun()
+        {
+            if (_baseCharacter.IsOnGround()) return false;
+
+            if (_baseCharacter.IsJumping())
+            {
+                _currentJumpDuration += Time.deltaTime;
+                if (_currentJumpDuration < _minJumpDuration) return false;
+            }
+
```

Comment:
how hard is it to check that the player has a forward momentum? Basically only allow wall running if the player is pressing {Forward Button} + {Left Button} || {Right Button}.

Since currently you can wall run backwards as well

Reply:
Yeah I think if I just call the input func and check if the y component is positive that should work, I'll test it